### PR TITLE
Use TString in AsciiOnlyNode and some refactoring 

### PR DESCRIPTION
**Pull request by wildmaples**

### CODE REVIEW

File: src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java [hunk 1]
```diff
@@ -453,7 +453,8 @@ protected Object begin(RubyMatchData matchData, int index,
             }
 
             final Rope matchDataSourceRope = strings.getRope(matchData.source);
-            if (multiByteCharacterProfile.profile(!singleByteOptimizableNode.execute(matchDataSourceRope))) {
+            if (multiByteCharacterProfile.profile(!singleByteOptimizableNode.execute(matchDataSourceRope,
+                    getContext().getEncodingManager().getRubyEncoding(matchDataSourceRope.encoding.getIndex())))) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java:457]:
```suggestion
                    strings.getEncoding(matchData.source)))) {
```
(same below)

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 2]
```diff
@@ -4397,12 +4404,12 @@ protected RubyBaseNodeWithExecute coercePatternToRope(RubyBaseNodeWithExecute pa
         }
 
         @Specialization(
-                guards = "singleByteOptimizableNode.execute(stringRope)")
+                guards = "singleByteOptimizableNode.execute(stringRope, " +
+                        "getContext().getEncodingManager().getRubyEncoding(stringRope.encoding.getIndex()))")
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java:4408]: This should be replaced by passing the RubyEncoding explicitly if possible (for efficiency & propagating the correct encoding).

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 3]
```diff
@@ -4431,7 +4438,8 @@ protected Object singleByteOptimizable(Rope stringRope, Rope patternRope, int of
 
         @TruffleBoundary
         @Specialization(
-                guards = "!singleByteOptimizableNode.execute(stringRope)")
+                guards = "!singleByteOptimizableNode.execute(stringRope, " +
+                        "getContext().getEncodingManager().getRubyEncoding(stringRope.encoding.getIndex()))")
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java:4442]: This should be replaced by passing the RubyEncoding explicitly if possible (for efficiency & propagating the correct encoding).

File: src/main/java/org/truffleruby/core/rope/RopeNodes.java [hunk 4]
```diff
@@ -897,21 +882,30 @@ public static SingleByteOptimizableNode create() {
...
+         * 
+         * @Cached AsciiOnlyNode asciiOnlyNode,
+         * 
+         * @Cached ConditionProfile asciiOnlyProfile,
+         * 
+         * @CachedLibrary(limit = "LIBSTRING_CACHE") RubyStringLibrary strings) { var encoding =
+         * strings.getEncoding(tString); final boolean asciiOnly = asciiOnlyNode.execute(tString, encoding);
+         * 
+         * if (asciiOnlyProfile.profile(asciiOnly)) { return true; } else { return encoding.jcoding.isSingleByte(); }
+         * } */
```
[the remark this answers is not in the document]
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/rope/RopeNodes.java:908]: It's leftover from when the branches were split. I'm sorry about that. The work on `SingleByteOptimizableNode` was bigger than we thought so it got moved out to another branch.

**Pull request merged**